### PR TITLE
[v1.1] Model Adapter code merge

### PR DESCRIPTION
**Pull request by XinyaoWa**

### CODE REVIEW

File: tests/cicd/TLKJenkinsfile [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            steps {
+                sh 'docker run --rm --name test-tlk-singlenode --privileged --network host --device=/dev/dri --shm-size="2g"  -v /mnt/DP_disk1/tlk:/home/vmagent/app/data  -v `pwd`:/home/vmagent/app -w /home/vmagent/app/test/cicd aidk-pytorch110 /bin/bash -c ". /home/vmagent/app/tests/cicd/jenkins_transfer_learning_unittest.sh"'
+            }
+        }
+        stage('Run transfer_learning_DDP_baseline_test') {
+            agent {
+               label 'sr613'
+            }
+            steps {
+               sh 'docker run -t -d --name test-tlk-multinode-baseline-r1 --privileged --network host --device=/dev/dri -v /mnt/DP_disk1/tlk:/home/vmagent/app/data  -v `pwd`:/home/vmagent/app -w /home/vmagent/app/AIDK/TransferLearningKit/src aidk-pytorch110 /bin/bash -c ". /home/vmagent/app/tests/cicd/jenkins_transfer_learning_basic_test_ddp.sh"'
```
Comment (zigzagcai) [inline tests/cicd/TLKJenkinsfile:38]: 1. Since ModelAdapter runs on PyTorch1.12 env, this image need to change to e2eaiok-pytorch112
2. Since the project name is e2eaiok, need to change aidk to e2eaiok.

File: e2eAIOK/ModelAdapter/example/Distiller_ResNet18_from_VIT_on_CIFAR100_save_logits.ipynb [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "7bc00450",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import sys\n",
+    "sys.path.append(\"/home/vmagent/app/TLK/frameworks.bigdata.AIDK/AIDK/\")\n",
```
Comment (zigzagcai) [inline e2eAIOK/ModelAdapter/example/Distiller_ResNet18_from_VIT_on_CIFAR100_save_logits.ipynb:59]: Since project name is e2eaiok, `frameworks.bigdata.AIDK/AIDK` need to be changed.

File: e2eAIOK/ModelAdapter/src/task/medical_segmentation/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
+## Environment Setup
+
+1. build docker image
+   ```
+   cd Dockerfile-ubuntu18.04 && docker build -t aidk-pytorch110 . -f DockerfilePytorch110 && cd .. && yes | docker container prune && yes | docker image prune
```
Comment (zigzagcai) [inline e2eAIOK/ModelAdapter/src/task/medical_segmentation/README.md:5]: please remove `aidk` typos and use e2eaiok

File: tests/cicd/jenkins_transfer_learning_basic_test.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+# init conda env
+eval "$('/opt/intel/oneapi/intelpython/latest/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
+# activate conda
+conda activate pytorch-1.10.0
```
Comment (zigzagcai) [inline tests/cicd/jenkins_transfer_learning_basic_test.sh:6]: If ModelAdapter runs with pytorch1.12.0 env, this env name need to be changed.

**Review by zigzagcai — CHANGES_REQUESTED** [review]

**Review by zigzagcai — APPROVED** [review]

File: .gitmodules [hunk 5]
```diff
@@ -25,3 +25,6 @@
 [submodule "modelzoo/third_party/IntelAI_models"]
 	path = modelzoo/third_party/IntelAI_models
 	url = https://github.com/IntelAI/models.git
+[submodule "e2eAIOK/ModelAdapter/src/task/medical_segmentation/third_party/nnUNet"]
+	path = e2eAIOK/ModelAdapter/src/task/medical_segmentation/third_party/nnUNet
+	url = https://github.com/MIC-DKFZ/nnUNet.git
```
Comment (Peach-He) [inline .gitmodules:30]: can we maitain all thirdparty code under same folder `modelzoo/third_party`?

File: tests/cicd/TLKJenkinsfile (hunk 1 above)
Comment (XinyaoWa) [reply tests/cicd/TLKJenkinsfile]: Updated

File: e2eAIOK/ModelAdapter/example/Distiller_ResNet18_from_VIT_on_CIFAR100_save_logits.ipynb (hunk 2 above)
Comment (XinyaoWa) [reply e2eAIOK/ModelAdapter/example/Distiller_ResNet18_from_VIT_on_CIFAR100_save_logits.ipynb]: Updated

File: e2eAIOK/ModelAdapter/src/task/medical_segmentation/README.md (hunk 3 above)
Comment (XinyaoWa) [reply e2eAIOK/ModelAdapter/src/task/medical_segmentation/README.md]: Updated

File: tests/cicd/jenkins_transfer_learning_basic_test.sh (hunk 4 above)
Comment (XinyaoWa) [reply tests/cicd/jenkins_transfer_learning_basic_test.sh]: Updated

**Comment by XinyaoWa** [comment]
> 

Thanks! We will add and check them later.

**Comment by XinyaoWa** [comment]
> unittest / CI tests for ModelAdapter need to be added to verify the functionality for model adapter. For CI tests, we can refer to `.github/workflows/integration_test_*.yml` For unittest, we can place UT under `tests/unittest/` folder, and run `pytest tests/unittest/` to verify.

Thanks! We will check and add them later.

File: .gitmodules (hunk 5 above)
Comment (XinyaoWa) [reply .gitmodules]: Updated

File: e2eAIOK/common/trainer/utils/utils.py [hunk 6]
```diff
@@ -414,4 +414,16 @@ def parse_activities(activity_str):
...
+    return profiler
+
+def tensor_near_equal(tensor1,tensor2,threshold = 1e-9):
+    ''' if tensor1 is near equal tensor2
+
+    :param tensor1: the first tensor
+    :param tensor2: the second tensor
+    :param threshold: threshold value
+    :return:
+    '''
```
Comment (xuechendi) [inline e2eAIOK/common/trainer/utils/utils.py:426]: cool! Thanks for adding dev docs here.

File: tests/cicd/jenkins_model_adapter_basic_test.sh [hunk 7]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+# init conda env
+eval "$('/opt/intel/oneapi/intelpython/latest/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
+# activate conda
+conda activate pytorch-1.12.0
+# run main.py
+cd /home/vmagent/app/e2eAIOK/e2eAIOK/ModelAdapter/src
```
Comment (xuechendi) [inline tests/cicd/jenkins_model_adapter_basic_test.sh:8]: @zigzagcai , is using 'e2eAIOK' correct? I checked with my docker env, should be '/home/vmagent/app/e2eaiok/e2eAIOK'

File: tests/cicd/ModelAdapterJenkinsfile [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+                sh 'docker run --rm --name test-ma-singlenode --privileged --network host --device=/dev/dri --shm-size="2g"  -v /mnt/DP_disk1/tlk:/home/vmagent/app/data  -v `pwd`:/home/vmagent/app -w /home/vmagent/app/test/cicd e2eaiok-pytorch120 /bin/bash -c ". /home/vmagent/app/tests/cicd/jenkins_model_adapter_unittest.sh"'
+            }
+        }
+        stage('Run model_adapter_DDP_baseline_test') {
+            agent {
+               label 'sr613'
+            }
+            steps {
+               sh 'docker run -t -d --name test-ma-multinode-baseline-r1 --privileged --network host --device=/dev/dri -v /mnt/DP_disk1/tlk:/home/vmagent/app/data  -v `pwd`:/home/vmagent/app -w /home/vmagent/app/e2eAIOK/e2eAIOK/ModelAdapter/src e2eaiok-pytorch120 /bin/bash -c ". /home/vmagent/app/tests/cicd/jenkins_model_adapter_basic_test_ddp.sh"'
+           }
```
Comment (xuechendi) [inline tests/cicd/ModelAdapterJenkinsfile:39]: @XinyaoWa @zigzagcai , I saw in other jenkins file, we are using /home/vmagent/app/e2eaiok/e2eAIOK instead of /home/vmagent/app/e2eAIOK/e2eAIOK, can this path also align with @XinyaoWa @zigzagcai , I saw in other jenkins file, we are using /home/vmagent/app/e2eaiok/e2eAIOK?

File: e2eAIOK/ModelAdapter/readme.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+- **Scheduler**: Scheduler defines how to schedule learning rate step-by-step, i.e. StepLR which decays learning periodically.
+
+<img src="./doc/imgs/arch.png" width="800px" />
+
+## Getting Started 
+### Installing
+1. Option 1: install with Docker (expected in v0.4)
+   ```bash
+   git clone https://github.com/intel/e2eAIOK.git
+   cd e2eAIOK; python scripts/start_e2eaiok_docker.py --backend  pytorch --dataset_path ./ --workers host1, host2, host3, host4
```
Comment (xuechendi) [inline e2eAIOK/ModelAdapter/readme.md:72]: @XinyaoWa , do you use pytorch as backend instead of pytorch112? Does this install guide verified ? 
This is how DeNas enabled: https://github.com/intel/e2eAIOK/tree/main/e2eAIOK/DeNas

Comment (xuechendi) [reply e2eAIOK/ModelAdapter/readme.md:72]: Meanwhile, I remembered you have some minor change to Dockerfile, didn't see in this PR

**Comment by xuechendi** [comment]
@XinyaoWa , would like to check with you the goal of this PR, if this is only for merging previous comment, please fix the jenkins path issue and I'll merge this PR firstly. And we can fix REAMDE later.

File: tests/cicd/jenkins_model_adapter_basic_test.sh (hunk 7 above)
Comment (XinyaoWa) [reply tests/cicd/jenkins_model_adapter_basic_test.sh]: Updated

File: tests/cicd/ModelAdapterJenkinsfile (hunk 8 above)
Comment (XinyaoWa) [reply tests/cicd/ModelAdapterJenkinsfile:39]: Updated

File: e2eAIOK/ModelAdapter/readme.md (hunk 9 above)
Comment (XinyaoWa) [reply e2eAIOK/ModelAdapter/readme.md]: Readme and Dockerfile updated.
